rosidl_generator_cpp/include/rosidl_generator_cpp/MessageTypeSupport.h

```diff
@@ -17,13 +17,22 @@
 #define __rosidl_generator_cpp__MessageTypeSupport__h__
 
 #include <rosidl_generator_c/message_type_support.h>
+#include <rosidl_generator_c/visibility_control.h>
 
 namespace rosidl_generator_cpp
 {
 
 template<typename T>
+ROSIDL_PUBLIC
 const rosidl_message_type_support_t * get_type_support_handle();
 
 }  // namespace rosidl_generator_cpp
 
+// Tail include the implementation for this function.
+// This include is provided by the implementation and therefore only
+// one version of this header will get included from the path.
+// This header maps the get_type_support_handle function to an implementation
+// specific version of that function.
+#include <rosidl_generator_cpp/get_type_support_handle_impl.hpp>
```

Comment:
I started out with what you are describing, but I changed to this for a few reasons. I'm willing to change it again, but I think some of the rational are compelling:

Being not includable without the correct include path set has actually helped to find places where this is used and the include path should be set, but isn't. I can't think of a situation where you would want to include this header, but didn't or couldn't setup the include path to something reasonable in order to get the implementation header.

The alternative strategy also results in non-includable files unless special include paths are set. Currently the consumer of this header includes just this header and must have the include paths set. In the other arrangement, consumers of this header always need to additionally include the implementation header, which can only be done with special include paths setup. So you include two headers instead of one with the same constraints.

The diff is pretty trivial, basically you remove this include line and comments and add an identical include line everywhere this header is included.